File: 659.split-array-into-consecutive-subsequences.cpp
/*
 * @lc app=leetcode id=659 lang=cpp
 *
 * [659] Split Array into Consecutive Subsequences
 */
#include <vector>
#include <queue>
using namespace std;
// @lc code=start
class Solution
{
public:
    bool isPossible(vector<int> &nums)
    {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pos_len_queue;
        for (auto n : nums)
        {
            while (!pos_len_queue.empty() && n > pos_len_queue.top().first + 1 && pos_len_queue.top().second >= 3)
            {
                pos_len_queue.pop();
            }
            if (pos_len_queue.empty())
            {
                pos_len_queue.push({n, 1});
            }
            else if (n == pos_len_queue.top().first)
            {
                pos_len_queue.push({n, 1});
            }
            else if (n == pos_len_queue.top().first + 1)
            {
                auto elem = pos_len_queue.top();
                pos_len_queue.pop();
                pos_len_queue.push({n, elem.second + 1});
            }
            else
            {
                return false;
            }
        }
        while (!pos_len_queue.empty())
        {
            if (pos_len_queue.top().second < 3)
            {
                return false;
            }
            pos_len_queue.pop();
        }
        return true;
    }
};
// @lc code=end

File: 659.split-array-into-consecutive-subsequences.java
import java.util.PriorityQueue;

/*
 * @lc app=leetcode id=659 lang=java
 *
 * [659] Split Array into Consecutive Subsequences
 */

// @lc code=start
class Solution {
    class Pair implements Comparable<Pair> {
        Integer first;
        Integer second;

        public Pair(Integer _first, Integer _second) {
            this.first = _first;
            this.second = _second;
        }

        @Override
        public int compareTo(Pair o) {
            return ((this.first - o.first) << 15) + this.second - o.second;
        }
    }

    public boolean isPossible(int[] nums) {
        PriorityQueue<Pair> pos_len_queue = new PriorityQueue<>();
        for (int n : nums) {
            while (!pos_len_queue.isEmpty() && n > pos_len_queue.peek().first + 1 && pos_len_queue.peek().second >= 3) {
                pos_len_queue.poll();
            }
            if (pos_len_queue.isEmpty()) {
                pos_len_queue.add(new Pair(n, 1));
            } else if (n == pos_len_queue.peek().first) {
                pos_len_queue.add(new Pair(n, 1));
            } else if (n == pos_len_queue.peek().first + 1) {
                var elem = pos_len_queue.poll();
                elem.first += 1;
                elem.second += 1;
                pos_len_queue.add(elem);
            } else {
                return false;
            }
        }
        while (!pos_len_queue.isEmpty()) {
            if (pos_len_queue.poll().second < 3) {
                return false;
            }
        }
        return true;
    }
}
// @lc code=end
